pkg/fqdn/messagehandler/message_handler.go

```diff
@@ -175,114 +192,28 @@ func (b *dnsRequestHandler) NotifyOnDNSMsg(
-		// updates for the response IPs. Due to G1 performing all the work
-		// first, G2 executes T4 also as a no-op and releases the msg back to the
-		// pod at T5 before G1 would at T6.
-		//
-		// We do not do a `defer unlock()` here, as we should release the lock before
-		// doing final bookkeeping.
-		mutexAcquireStart := time.Now()
-		b.nameManager.LockName(qname)
-
-		if d := time.Since(mutexAcquireStart); d >= option.Config.DNSProxyLockTimeout {
-			b.logger.Warn(fmt.Sprintf("Name lock acquisition time took longer than expected. Potentially too many parallel DNS requests being processed, consider adjusting --%s and/or --%s", option.DNSProxyLockCount, option.DNSProxyLockTimeout),
+		if err := h.UpdateOnDNSMsg(lookupTime, ep, qname, responseIPs, int(TTL), stat); err != nil {
+			h.logger.Error("failed to update DNS message",
+				logfields.Error, err,
 				logfields.DNSName, qname,
-				logfields.Duration, d,
-				logfields.Expected, option.Config.DNSProxyLockTimeout,
+				logfields.IPAddrs, responseIPs,
 			)
 		}
-
-		b.logger.Debug("Recording DNS lookup in endpoint specific cache", logfields.EndpointID, ep.ID)
-
-		// This must happen before the NameManager update below, to ensure that
-		// this data is included in the serialized Endpoint object.
-		// We also need to add to the cache before we purge any matching zombies
-		// because they are locked separately and we want to keep the allowed IPs
-		// consistent if a regeneration happens between the two steps. If an update
-		// doesn't happen in the case, we play it safe and don't purge the zombie
-		// in case of races.
-		if updated := ep.DNSHistory.Update(lookupTime, qname, responseIPs, int(TTL)); updated {
-			ep.DNSZombies.ForceExpireByNameIP(lookupTime, qname, responseIPs...)
-			ep.SyncEndpointHeaderFile()
-		}
-
-		b.logger.Debug("Updating DNS name in cache from response to query",
-			logfields.DNSName, qname,
-			logfields.IPAddrs, responseIPs,
-		)
-
-		updateCtx, updateCancel := context.WithTimeout(b.ctx, option.Config.FQDNProxyResponseMaxDelay)
-		defer updateCancel()
-		updateStart := time.Now()
-
-		dpUpdates := b.nameManager.UpdateGenerateDNS(updateCtx, lookupTime, qname, &fqdn.DNSIPRecords{
-			IPs: responseIPs,
-			TTL: int(TTL),
-		})
-
-		stat.PolicyGenerationTime.End(true)
-		stat.DataplaneTime.Start()
-
-		if err := <-dpUpdates; err != nil {
-			b.logger.Warn("Timed out waiting for datapath updates of FQDN IP information; returning response. Consider increasing --tofqdns-proxy-response-max-delay if this keeps happening.")
-			metrics.ProxyDatapathUpdateTimeout.Inc()
-		}
-
-		// Policy updates for this name have been pushed out; we can release the lock.
-		b.nameManager.UnlockName(qname)
-
-		b.logger.Debug("Waited for endpoints to regenerate due to a DNS response",
-			logfields.Duration, time.Since(updateStart),
-			logfields.EndpointID, ep.GetID(),
-			logfields.DNSName, qname,
-		)
-
 		endMetric()
```

Comment:
I think we should split out some logic out of `endMetric()` too. In particular, I think we should move the `stat.DataplaneTime.End(true)` into a defer statement into `UpdateOnDNSMsg`, since `UpdateOnDNSMsg` is also the one which starts it.

Reply:
Make sense, updated it.

Reply:
Make sense, updated it.